### PR TITLE
More generic extrinsic pool

### PR DESCRIPTION
**Pull request by arkpar**
* Moved most of the common logic from polkadot transaction pool to substrate.
* Extrinsic hash type can now be different from the block hash type.

### CODE REVIEW

File: substrate/service/src/lib.rs [hunk 1]
```diff
@@ -241,13 +248,13 @@ impl<Components> Service<Components>
...
 					on_connect: Box::new(move || {
 						telemetry!("system.connected";
-							"name" => name.clone(),
-							"implementation" => impl_name.clone(),
-							"version" => version.clone(),
-							"config" => "",
-							"chain" => chain_name.clone(),
-							"pubkey" => &pubkey,
-						);
+								   "name" => name.clone(),
```
Comment (pepyakin) [inline substrate/service/src/lib.rs]: Looks like there are some spaces sneaked in.

File: substrate/service/src/lib.rs [hunk 2]
```diff
@@ -343,3 +350,76 @@ impl substrate_rpc::system::SystemApi for RpcConfig {
...
+
+		let encoded = transaction.encode();
+		if let Some(uxt) = Decode::decode(&mut &encoded[..]) {
+			let best_block_id = self.best_block_id()?;
+			match self.pool.submit_one(&best_block_id, uxt) {
+				Ok(xt) => Some(*xt.hash()),
+				Err(e) => match e.into_pool_error() {
+					Ok(e) => match e.kind() {
+						extrinsic_pool::ErrorKind::AlreadyImported(hash) =>
+							Some(::std::str::FromStr::from_str(&hash).map_err(|_| {}).expect("Hash string is always valid")),
```
Comment (pepyakin) [inline substrate/service/src/lib.rs]: Can we wrap this line ?

File: substrate/service/src/lib.rs [hunk 3]
```diff
@@ -343,3 +350,76 @@ impl substrate_rpc::system::SystemApi for RpcConfig {
...
+						_ => {
+							debug!("Error adding transaction to the pool: {:?}", e);
+							None
+						},
+					},
+					Err(e) => {
+							debug!("Error converting pool error: {:?}", e);
```
Comment (pepyakin) [inline substrate/service/src/lib.rs]: The indentation looks a bit off here

File: substrate/service/src/lib.rs (hunk 1 above)
Comment (pepyakin) [reply substrate/service/src/lib.rs]: Why this was indentend anyway?

File: substrate/service/src/lib.rs [hunk 4]
```diff
@@ -209,14 +217,13 @@ impl<Components> Service<Components>
...
-
-				rpc::rpc_handler::<ComponentBlock<Components>, _, _, _, _, _>(
+				let author = rpc::apis::author::Author::new(client.clone(), extrinsic_pool.clone(), task_executor.clone());
+				rpc::rpc_handler::<ComponentBlock<Components>, ComponentExHash<Components>, _, _, _, _, _>(
 					state,
 					chain,
 					author,
 					rpc_config.clone(),
-				)
+					)
```
Comment (pepyakin) [inline substrate/service/src/lib.rs]: inconsistent indentation

File: substrate/rpc/src/author/tests.rs [hunk 5]
```diff
@@ -16,73 +16,79 @@
...
+	Readiness, ExtrinsicFor, VerifiedFor};
+use test_client::runtime::{Block, Extrinsic, Transfer};
 use test_client;
 use tokio::runtime;
+use runtime_primitives::generic::BlockId;
 
-type Extrinsic = u64;
-type Hash = u64;
+#[derive(Clone, Debug)]
+pub struct Verified(u64, u64); // sender, nonce
```
Comment (pepyakin) [inline substrate/rpc/src/author/tests.rs]: Wouldn't it be more readable if it were a normal struct instead of a tuple?

**Comment by gavofyork** [comment]
looks reasonable.

**Pull request merged**